Applications/spire/source/ui/dropdown_menu_list.cpp

```diff
@@ -0,0 +1,184 @@
+    }
+
+    QScrollBar::sub-line:vertical {
+      background: none;
+      border: none;
+    }
+
+    QScrollBar::add-line:vertical {
+      background: none;
+      border: none;
+    })").arg(scale_width(15)).arg(scale_width(1)).arg(scale_width(2)));
+  layout->addWidget(m_scroll_area);
+  m_list_widget = new QWidget(m_scroll_area);
+  auto list_layout = new QVBoxLayout(m_list_widget);
+  list_layout->setContentsMargins({});
+  list_layout->setSpacing(0);
+  for(auto& item : items) {
+    auto i = new DropdownMenuItem(item, m_list_widget);
+    i->connect_selected_signal([=] (auto& t) { on_select(t); });
+    list_layout->addWidget(i);
+  }
+  m_list_widget->setStyleSheet("background-color: #FFFFFF;");
+  m_scroll_area->setWidget(m_list_widget);
+  parent->installEventFilter(this);
+}
+
+void DropdownMenuList::set_items(const std::vector<QString>& items) {
+  while(auto item = m_list_widget->layout()->takeAt(0)) {
+    delete item->widget();
+    delete item;
+  }
+  for(auto& item : items) {
+    auto i = new DropdownMenuItem(item, m_list_widget);
+    i->connect_selected_signal([=] (auto& t) { on_select(t); });
+    m_list_widget->layout()->addWidget(i);
+  }
+}
+
+QString DropdownMenuList::get_next(const QString& text) {
+  auto num_items = m_list_widget->layout()->count();
+  auto index = (get_index(text) + 1) % num_items;
+  return static_cast<DropdownMenuItem*>(
+    m_list_widget->layout()->itemAt(index)->widget())->text();
+}
+
+QString DropdownMenuList::get_previous(const QString& text) {
+  auto num_items = m_list_widget->layout()->count();
+  auto index = (get_index(text) - 1) % num_items;
+  if(index < 0) {
+    index = num_items - 1;
+  }
+  return static_cast<DropdownMenuItem*>(
+    m_list_widget->layout()->itemAt(index)->widget())->text();
+}
+
+connection DropdownMenuList::connect_selected_signal(
+    const SelectedSignal::slot_type& slot) const {
+  return m_selected_signal.connect(slot);
+}
+
+bool DropdownMenuList::eventFilter(QObject* object, QEvent* event) {
+  if(object == parent()) {
+    if(event->type() == QEvent::KeyPress && isVisible()) {
+      auto e = static_cast<QKeyEvent*>(event);
+      if(e->key() == Qt::Key_Tab || e->key() == Qt::Key_Down) {
+        focus_next();
+        return true;
+      } else if((e->key() & Qt::Key_Tab &&
+          e->modifiers() & Qt::ShiftModifier) || e->key() == Qt::Key_Up) {
+        focus_previous();
+        return true;
+      } else if(e->key() == Qt::Key_Enter || e->key() == Qt::Key_Return) {
+        on_select(static_cast<DropdownMenuItem*>(m_list_widget->layout()->
+          itemAt(m_highlight_index)->widget())->text());
+        return true;
+      } else if(e->key() == Qt::Key_Escape) {
+        close();
+      }
+    }
+  }
+  return false;
+}
+
+void DropdownMenuList::showEvent(QShowEvent* event) {
+  m_highlight_index = -1;
+  for(auto i = 0; i < m_list_widget->layout()->count(); ++i) {
+    auto w = m_list_widget->layout()->itemAt(i)->widget();
```

Comment:
```auto widget = ...``` or ```item_widget```, but avoid using one letter names except for iterators or mathematical operations.